### PR TITLE
Consolidate `ConstantFolding` and `SimplifyExpression`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Builds on https://github.com/apache/arrow-datafusion/pull/1374

Re https://github.com/apache/arrow-datafusion/issues/1160 (more sophisticated expression simplification) and #1316  don't change names during constant folding

 # Rationale for this change
`ConstantFolding` and `SimplifyExpression` both simplify expressions, in slightly different ways, which is confusing and redundant. 

It also means bug fixes such as #1319 didn't get applied to `SimplifyExpression`.

For example, on master (not the column name is `column1`, due to the simplification that has occured):

```
❯ select column1 * 1 from (values (1, NULL), (2, 2), (3,3)) as sq;
+---------+
| column1 |
+---------+
| 1       |
| 2       |
| 3       |
+---------+
```

On this branch the column is correctly named:
```sql
❯ select column1 * 1 from (values (1, NULL), (2, 2), (3,3)) as sq;
+-----------------------+
| sq.column1 * Int64(1) |
+-----------------------+
| 1                     |
| 2                     |
| 3                     |
+-----------------------+
```

# What changes are included in this PR?
1. Remove independent `ConstantFolding` optimizer pass by compiling `ConstantFolding` into `SimplifyExpression` optimizer pass
2. Move tests

Note:
I plan more PRs to combine `simplify` and `Simplifier` logic but to keep reviews easier I am going to do that in another PR

### CODE REVIEW

File: datafusion/src/optimizer/simplify_expressions.rs [hunk 1]
```diff
@@ -284,9 +274,62 @@ impl OptimizerRule for SimplifyExpressions {
...
-        _execution_props: &ExecutionProps,
+        execution_props: &ExecutionProps,
     ) -> Result<LogicalPlan> {
-        optimize(plan)
+        // We need to pass down the all schemas within the plan tree to `optimize_expr` in order to
+        // to evaluate expression types. For example, a projection plan's schema will only include
+        // projected columns. With just the projected schema, it's not possible to infer types for
+        // expressions that references non-projected columns within the same project plan or its
+        // children plans.
+        let mut simplifier =
```
Comment (alamb) [inline datafusion/src/optimizer/simplify_expressions.rs:284]: this logic is moved from ConstantFolding

File: datafusion/src/optimizer/simplify_expressions.rs [hunk 2]
```diff
@@ -284,9 +274,62 @@ impl OptimizerRule for SimplifyExpressions {
...
+        let expr = plan
+            .expressions()
+            .into_iter()
+            .map(|e| {
+                // We need to keep original expression name, if any.
+                // Constant folding should not change expression name.
+                let name = &e.name(plan.schema());
+
+                // TODO combine simplify into Simplifier
+                let e = simplify(&e);
```
Comment (alamb) [inline datafusion/src/optimizer/simplify_expressions.rs:305]: Here is the call into the existing `simplify` function that is in the `SimplifyExpressions` pass (originally written by @jgoday)

File: datafusion/src/optimizer/simplify_expressions.rs [hunk 3]
```diff
@@ -284,9 +274,62 @@ impl OptimizerRule for SimplifyExpressions {
...
+                // evaluation)
+                let new_e = e
+                    // fold constants and then simplify
+                    .rewrite(&mut const_evaluator)?
+                    .rewrite(&mut simplifier)?;
+
+                let new_name = &new_e.name(plan.schema());
+
+                // TODO simplify this logic
+                if let (Ok(expr_name), Ok(new_expr_name)) = (name, new_name) {
```
Comment (alamb) [inline datafusion/src/optimizer/simplify_expressions.rs:320]: Note that now the  fix for alias names (introduced by @viirya in https://github.com/apache/arrow-datafusion/pull/1319) is applied to the result of `simplify` as well -- previously it was not

File: datafusion/src/optimizer/simplify_expressions.rs [hunk 4]
```diff
@@ -1562,4 +1556,388 @@ mod tests {
             .unwrap(),
         )
     }
+
```
Comment (alamb) [inline datafusion/src/optimizer/simplify_expressions.rs:1559]: these tests are just moved from `constant_folding.rs`

**Comment by alamb** [comment]
The new tpch query test added in #1368 by @viirya  is already coming in handy to prevent regressions -- it found a [problem](https://github.com/apache/arrow-datafusion/runs/4366849992?check_suite_focus=true) with an earlier version of this pr where simplification wasn't done until after the other optimizations. ❤️

File: datafusion/src/execution/context.rs [hunk 5]
```diff
@@ -896,12 +895,13 @@ impl Default for ExecutionConfig {
             target_partitions: num_cpus::get(),
             batch_size: 8192,
             optimizers: vec![
-                Arc::new(ConstantFolding::new()),
+                // Simplify expressions first to maximize the chance
+                // of applying other optimizations
+                Arc::new(SimplifyExpressions::new()),
```
Comment (alamb) [inline datafusion/src/execution/context.rs:900]: This change has the additional benefit of allowing even more filters to be pushed down (e.g. `column * 1 < 5` which is simplified by the code in `simpify` to `column < 5` can now also be pushed down

**Review by houqp — APPROVED** [review]
nice simplification and tests :)

**Pull request merged**